hathor/p2p/sync_checkpoints.py

```diff
@@ -0,0 +1,213 @@
+# Copyright 2021 Hathor Labs
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from itertools import chain
+from typing import TYPE_CHECKING, Dict, List, Optional, Tuple
+
+from structlog import get_logger
+
+from hathor.checkpoint import Checkpoint
+from hathor.transaction.block import Block
+from hathor.transaction.storage.exceptions import TransactionDoesNotExist
+
+if TYPE_CHECKING:
+    from hathor.manager import HathorManager  # noqa: F401
+    from hathor.p2p.node_sync_v2 import NodeBlockSync  # noqa: F401
+
+logger = get_logger()
+
+
+class SyncCheckpoint:
+    """ It is a central manager of the sync between checkpoints among all peers.
+    """
+
+    def __init__(self, manager: 'HathorManager'):
+        # All peers that have all the checkpoints to download
+        self.peers_to_request: List['NodeBlockSync'] = []
+
+        # All peers that we are connected but don't have all the checkpoints
+        self.incomplete_peers: List['NodeBlockSync'] = []
+
+        # All peers that we tried to download but they sent wrong blocks
+        self.banned_peers: List['NodeBlockSync'] = []
+
+        # If the object is running sync
+        self.is_running: bool = False
+
+        # HathorManager object to get checkpoints and storage
+        self.manager: 'HathorManager' = manager
+
+        # All checkpoints that still need to sync
+        self.checkpoints_to_sync: List[Checkpoint] = []
+
+        # Previous checkpoints map
+        self.previous_checkpoints: Dict[Checkpoint, Checkpoint] = {}
+
+        # The peer that is syncing (the one we are downloading the blocks from)
+        self.peer_syncing = None
+
+        # If set to true next run_sync_transactions will be skipped
+        self.should_skip_sync_tx = False
+
+        # Create logger with context
+        self.log = logger.new()
+
+    def start(self):
+        if self.is_running:
+            self.log.debug('alreay running, not starting new one')
+
+        self.log.debug('start sync')
+        checkpoints = self.manager.checkpoints
+        bestblock = self.manager.tx_storage.get_best_block()
+        meta = bestblock.get_metadata()
+
+        # Fill the previous checkpoints map
+        it_cps = iter(checkpoints)
```

Comment:
Should we add the checkpoint index to the `Checkpoint` namedtuple? It seems useful.

Reply:
I think `Checkpoint` should be lightweight (in general NamedTuples should be lightweight IMO). Maybe we could make something like a linked list entry to easily have the previous checkpoint, but this seems very minor, I don't think we will ever need to worry about the checkpoint list being too large (at most we'll make it sparser for older heights).

Reply:
I think `Checkpoint` should be lightweight (in general NamedTuples should be lightweight IMO). Maybe we could make something like a linked list entry to easily have the previous checkpoint, but this seems very minor, I don't think we will ever need to worry about the checkpoint list being too large (at most we'll make it sparser for older heights).

Reply:
I agree.